elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/HttpClient.scala

```diff
@@ -59,24 +56,14 @@ trait HttpClient extends Logging {
   // Executes the given request type T, and returns a Future of Response[U] where U is particular to the request type.
   // For example a search request will return a Future[Response[SearchResponse]].
   // The returned Response is an ADT
-  def execute[T, U](request: T)(implicit exec: HttpExecutable[T, U]): Future[Either[RequestFailure, RequestSuccess[U]]] = {
-    val p = Promise[Either[RequestFailure, RequestSuccess[U]]]()
-    val f = exec.execute(client, request)
-    f.onComplete {
-      case Success(r) =>
-        try exec.responseHandler.handle(r) match {
-          case Right(u) =>
-            val success = RequestSuccess(r.statusCode, r.entity.map(_.content), r.headers, u)
-            p.trySuccess(Right(success))
-          case Left(error) =>
-            val failure = RequestFailure(r.statusCode, r.entity.map(_.content), r.headers, error)
-            p.trySuccess(Left(failure))
-        } catch {
-          case t: Throwable => p.tryFailure(t)
-        }
-      case Failure(t) => p.tryFailure(t)
-    }
-    p.future
+  def execute[T, U](request: T)(implicit exec: HttpExecutable[T, U], executionContext: ExecutionContext): Future[Either[RequestFailure, RequestSuccess[U]]] = {
+    exec.execute(client, request)
+      .map(r => exec.responseHandler.handle(r) match {
```

Comment:
I don't think this handles the case where the response handler throws an exception, which I'd like to see materialized as a Future.failure like before.

Reply:
It wouldn't be safe in `onComplete`, but I don't think exceptions in `map` escape the context of Future.

```scala
val f = Future.successful("not a number").map(_.toInt)
f.failed.map(_ => println("failed future"))
```

results in a proper Future.failed.

https://docs.scala-lang.org/overviews/core/futures.html

> If the mapping function throws an exception the future is completed with that exception.

I think the 2 pieces of code are functionally equivalent, unless I'm forgetting to take something into account.

Reply:
Ok